third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+    // for map default insert
+    FatPtrAttrs() = default;
+    bool canNarrow = false;
+    llvm::SmallDenseMap<StringRef, Attribute> attributes;
+    friend bool operator==(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return lhs.canNarrow == rhs.canNarrow && lhs.attributes == rhs.attributes;
+    }
+    friend bool operator!=(const FatPtrAttrs &lhs, const FatPtrAttrs &rhs) {
+      return !(lhs == rhs);
+    }
+  };
+  using KeyT = std::pair<Value, Value>;
+  using ValueT = FatPtrAttrs;
+  using DenseMapT = DenseMap<KeyT, ValueT>;
+  void collectFatPointerAttributes(const KeyT &k);
+  ValueT &operator[](const KeyT &k) {
+    if (!pointerAttrs.contains(k))
+      collectFatPointerAttributes(k);
+    return pointerAttrs[k];
+  }
+  ValueT &operator[](KeyT &&k) {
+    if (!pointerAttrs.contains(k))
+      collectFatPointerAttributes(k);
+    return pointerAttrs[k];
+  }
+  template <typename T>
+  using const_arg_type_t = typename llvm::const_pointer_or_const_ref<T>::type;
+  const ValueT &at(const_arg_type_t<KeyT> k) const {
+    // this is redundant - DenseMap will assert the same thing - but better to
+    // have our own message
+    assert(pointerAttrs.contains(k) &&
+           "expected fatPtrs to contain remapped fat pointer");
+    return pointerAttrs.at(k);
+  }
+  const bool contains(const KeyT &k) { return pointerAttrs.contains(k); }
+
+private:
+  DenseMapT pointerAttrs;
+};
+
+// TODO(max): this is not a good way to do this...
+void FatPointers::collectFatPointerAttributes(const KeyT &k) {
+  auto [base, offset] = k;
+  // If it is the i-th block argument, then look if the operation defined some
+  // _argi attribute and add it to the fat pointer attributes
+  if (auto arg = dyn_cast<BlockArgument>(base)) {
+    // If the value is a block parameter, the operation can specify
+    // an attribute for the given parameter by using `tt.property_argi`
+    // where `argi` refers to the arg number of the given parameter.
+    // So we need to iterate through the property, find the right one
+    // and push the property onto the pointers attributes.
+    auto op = arg.getOwner()->getParentOp();
+    for (NamedAttribute namedAttr : op->getAttrs()) {
+      StringAttr attrName = namedAttr.getName();
+      std::string argSuffix =
+          llvm::formatv("_arg{0}", arg.getArgNumber()).str();
+      if (!attrName.strref().ends_with(argSuffix))
+        continue;
+
+      auto newAttrName = attrName.strref().drop_back(argSuffix.size());
+      pointerAttrs[k].attributes[newAttrName] = namedAttr.getValue();
+      // Propagate the argument to the offset if it is also a block
+      // argument
+      if (auto offsetArg = dyn_cast<BlockArgument>(offset))
+        op->setAttr(
+            llvm::formatv("{0}_arg{1}", newAttrName, offsetArg.getArgNumber())
+                .str(),
+            namedAttr.getValue());
+    }
+    return;
+  }
+
+  // TODO(max): this doesn't make sense - ops have all sorts of dialect
```

Comment:
Was this to be removed?

Reply:
It's fine to keep some TODO here but we should remove the commented out code--they are disconnected and confusing to be here.